maven-resolver-provider/src/main/java/org/apache/maven/repository/internal/PluginsMetadataGenerator.java

```diff
@@ -124,7 +128,46 @@ public Collection<? extends Metadata> finish( Collection<? extends Artifact> art
                 }
             }
         }
-
         return plugins.values();
     }
+
+    private PluginInfo extractPluginInfo( Artifact artifact )
+    {
+        // sanity: jar, no classifier and file exists
+        if ( artifact != null
+                && "jar".equals( artifact.getExtension() )
+                && "".equals( artifact.getClassifier() )
+                && artifact.getFile() != null )
+        {
+            Path artifactPath = artifact.getFile().toPath();
+            if ( Files.isRegularFile( artifactPath ) )
+            {
+                try ( JarFile artifactJar = new JarFile( artifactPath.toFile(), false ) )
+                {
+                    ZipEntry pluginDescriptorEntry = artifactJar.getEntry( PLUGIN_DESCRIPTOR_LOCATION );
+
+                    if ( pluginDescriptorEntry != null )
+                    {
+                        InputStream is = artifactJar.getInputStream( pluginDescriptorEntry );
+
+                        Reader reader = ReaderFactory.newXmlReader( is );
+
+                        PluginDescriptor pluginDescriptor =
+                                pluginDescriptorBuilder.build( reader, artifactPath.toAbsolutePath().toString() );
+
+                        return new PluginInfo(
+                                pluginDescriptor.getGroupId(),
+                                pluginDescriptor.getArtifactId(),
+                                pluginDescriptor.getGoalPrefix(),
+                                pluginDescriptor.getName() );
+                    }
+                }
+                catch ( Exception e )
+                {
+                    // ignore it for now
```

Comment:
This is too broad, but for example maven's ArtifactDeployerTest and ArtifactInstallerTest (maven-compat) uses invalid JAR for tests.... :sob: